[bug] Follow request from GoToSocial causes nil pointer exception

Here's the follow request as it arrives in the inbox:

```
 {"@context":"https://www.w3.org/ns/activitystreams","actor":"https://goblin.technology/users/tobi","id":"https://goblin.technology/users/tobi/follow/017E3ZYTD3V02DYA69E6D42S6T","object":"https://live.superseriousbusiness.org/federation/user/streamer","to":"https://live.superseriousbusiness.org/federation/user/streamer","type":"Follow"} 
```

And here's the panic:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x173ccd4]

goroutine 74 [running]:
github.com/owncast/owncast/activitypub/apmodels.MakeActorFromExernalAPEntity({0x7f18e3e6aad8, 0xc000208600})
	/build/activitypub/apmodels/actor.go:84 +0x274
github.com/owncast/owncast/activitypub/resolvers.GetResolvedActorFromIRI.func1({0x19b0a00?, 0x0?}, {0x24110e8?, 0xc000208600})
	/build/activitypub/resolvers/resolve.go:131 +0x4f
github.com/go-fed/activity/streams.JSONResolver.Resolve.func1({0xc0005200aa, 0x6})
	/go/pkg/mod/github.com/owncast/activity@v1.0.1-0.20211229051252-7821289d4026/streams/gen_json_resolver.go:691 +0x3d13
github.com/go-fed/activity/streams.JSONResolver.Resolve({{0xc0005bd380?, 0x2fc?, 0xc00040ba88?}}, {0x23ee618, 0xc000036060}, 0xc00082c120)
	/go/pkg/mod/github.com/owncast/activity@v1.0.1-0.20211229051252-7821289d4026/streams/gen_json_resolver.go:888 +0x171
github.com/owncast/owncast/activitypub/resolvers.Resolve({0x23ee618, 0xc000036060}, {0xc000846000, 0x2fc, 0x380}, {0xc0005bd380?, 0x0?, 0x0?})
	/build/activitypub/resolvers/resolve.go:35 +0x159
github.com/owncast/owncast/activitypub/resolvers.ResolveIRI({0x23ee618, 0xc000036060}, {0xc0006de1c0, 0x2d}, {0xc0005bd380, 0x3, 0x3})
	/build/activitypub/resolvers/resolve.go:72 +0x245
github.com/owncast/owncast/activitypub/resolvers.GetResolvedActorFromIRI({0xc0006de1c0, 0x2d})
	/build/activitypub/resolvers/resolve.go:157 +0x245
github.com/owncast/owncast/activitypub/inbox.Verify(0xc0003c5000)
	/build/activitypub/inbox/worker.go:74 +0x358
github.com/owncast/owncast/activitypub/inbox.handle({0xc0003c5000, {0xc0001ef5a6, 0x8}, {0xc0006e4000, 0x151, 0x200}})
	/build/activitypub/inbox/worker.go:23 +0x65
github.com/owncast/owncast/activitypub/inbox.worker(0x0?, 0x0?)
	/build/activitypub/inbox/workerpool.go:40 +0x108
created by github.com/owncast/owncast/activitypub/inbox.InitInboxWorkerPool
	/build/activitypub/inbox/workerpool.go:26 +0x51
```

This is with the latest version of owncast from develop. I'm gonna try and track this one down myself :)

--- a/activitypub/inbox/worker.go
+++ b/activitypub/inbox/worker.go
@@ -119,11 +119,11 @@ func Verify(request *http.Request) (bool, error) {
 	triedAlgos := make(map[httpsig.Algorithm]error)
 	for _, algorithm := range algos {
 		if _, tried := triedAlgos[algorithm]; !tried {
-			if err := verifier.Verify(parsedKey, algorithm); err == nil {
+			err := verifier.Verify(parsedKey, algorithm)
+			if err == nil {
 				return true, nil
-			} else {
-				triedAlgos[algorithm] = err
 			}
+			triedAlgos[algorithm] = err
 		}
 	}
 

--- a/activitypub/inbox/worker.go
+++ b/activitypub/inbox/worker.go
@@ -108,14 +108,21 @@ func Verify(request *http.Request) (bool, error) {
 		return false, errors.Wrap(err, "failed to parse DER encoded public key")
 	}
 
-	var algorithm httpsig.Algorithm = httpsig.Algorithm(algorithmString)
-
-	// The verifier will verify the Digest in addition to the HTTP signature
-	if err := verifier.Verify(parsedKey, algorithm); err != nil {
-		return false, errors.Wrap(err, algorithmString+" http signature verification error for: "+pubKeyID.String())
+	algos := []httpsig.Algorithm{
+		httpsig.RSA_SHA256, // most common algorithm first
+		httpsig.RSA_SHA512,
+		httpsig.ED25519,
+		httpsig.Algorithm(algorithmString),
 	}
 
-	return true, nil
+	// The verifier will verify the Digest in addition to the HTTP signature
+	for _, algorithm := range algos {
+		if err := verifier.Verify(parsedKey, algorithm); err == nil {
+			return true, nil
+		}
+	}
+
+	return false, errors.Wrap(err, algorithmString+" http signature verification error for: "+pubKeyID.String())
 }
 
 func isBlockedDomain(domain string) bool {

--- a/activitypub/inbox/worker.go
+++ b/activitypub/inbox/worker.go
@@ -115,9 +115,14 @@ func Verify(request *http.Request) (bool, error) {
 	}
 
 	// The verifier will verify the Digest in addition to the HTTP signature
+	triedAlgos := make(map[httpsig.Algorithm]error)
 	for _, algorithm := range algos {
-		if err := verifier.Verify(parsedKey, algorithm); err == nil {
-			return true, nil
+		if _, tried := triedAlgos[algorithm]; !tried {
+			if err := verifier.Verify(parsedKey, algorithm); err == nil {
+				return true, nil
+			} else {
+				triedAlgos[algorithm] = err
+			}
 		}
 	}
 

--- a/activitypub/inbox/worker.go
+++ b/activitypub/inbox/worker.go
@@ -71,26 +71,31 @@ func Verify(request *http.Request) (bool, error) {
 		return false, errors.New("Unable to determine algorithm to verify request")
 	}
 
-	actor, err := resolvers.GetResolvedActorFromIRI(pubKeyID.String())
+	publicKey, err := resolvers.GetResolvedPublicKeyFromIRI(pubKeyID.String())
 	if err != nil {
 		return false, errors.Wrap(err, "failed to resolve actor from IRI to fetch key")
 	}
 
-	if actor.ActorIri == nil {
-		return false, errors.New("actor IRI is empty")
+	var publicKeyActorIRI *url.URL
+	if ownerProp := publicKey.GetW3IDSecurityV1Owner(); ownerProp != nil {
+		publicKeyActorIRI = ownerProp.Get()
+	}
+
+	if publicKeyActorIRI == nil {
+		return false, errors.New("public key owner IRI is empty")
 	}
 
 	// Test to see if the actor is in the list of blocked federated domains.
-	if isBlockedDomain(actor.ActorIri.Hostname()) {
+	if isBlockedDomain(publicKeyActorIRI.Hostname()) {
 		return false, errors.New("domain is blocked")
 	}
 
 	// If actor is specifically blocked, then fail validation.
-	if blocked, err := isBlockedActor(actor.ActorIri); err != nil || blocked {
+	if blocked, err := isBlockedActor(publicKeyActorIRI); err != nil || blocked {
 		return false, err
 	}
 
-	key := actor.W3IDSecurityV1PublicKey.Begin().Get().GetW3IDSecurityV1PublicKeyPem().Get()
+	key := publicKey.GetW3IDSecurityV1PublicKeyPem().Get()
 	block, _ := pem.Decode([]byte(key))
 	if block == nil {
 		log.Errorln("failed to parse PEM block containing the public key")

--- a/activitypub/inbox/worker.go
+++ b/activitypub/inbox/worker.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"crypto/x509"
 	"encoding/pem"
+	"fmt"
 	"net/http"
 	"net/url"
 	"strings"
@@ -126,7 +127,7 @@ func Verify(request *http.Request) (bool, error) {
 		}
 	}
 
-	return false, errors.Wrap(err, algorithmString+" http signature verification error for: "+pubKeyID.String())
+	return false, fmt.Errorf("http signature verification error(s) for: %s: %+v", pubKeyID.String(), triedAlgos)
 }
 
 func isBlockedDomain(domain string) bool {

--- a/activitypub/inbox/worker.go
+++ b/activitypub/inbox/worker.go
@@ -109,10 +109,9 @@ func Verify(request *http.Request) (bool, error) {
 	}
 
 	algos := []httpsig.Algorithm{
-		httpsig.RSA_SHA256, // most common algorithm first
+		httpsig.Algorithm(algorithmString), // try stated algorithm first then other common algorithms
+		httpsig.RSA_SHA256,                 // <- used by almost all fedi software
 		httpsig.RSA_SHA512,
-		httpsig.ED25519,
-		httpsig.Algorithm(algorithmString),
 	}
 
 	// The verifier will verify the Digest in addition to the HTTP signature

--- a/activitypub/resolvers/resolve.go
+++ b/activitypub/resolvers/resolve.go
@@ -122,6 +122,72 @@ func GetResolvedActorFromActorProperty(actor vocab.ActivityStreamsActorProperty)
 	return apActor, err
 }
 
+// GetResolvedPublicKeyFromIRI will resolve a publicKey IRI string to a vocab.W3IDSecurityV1PublicKey.
+func GetResolvedPublicKeyFromIRI(publicKeyIRI string) (vocab.W3IDSecurityV1PublicKey, error) {
+	var err error
+	var pubkey vocab.W3IDSecurityV1PublicKey
+	resolved := false
+
+	personCallback := func(c context.Context, person vocab.ActivityStreamsPerson) error {
+		if pkProp := person.GetW3IDSecurityV1PublicKey(); pkProp != nil {
+			for iter := pkProp.Begin(); iter != pkProp.End(); iter = iter.Next() {
+				if iter.IsW3IDSecurityV1PublicKey() {
+					pubkey = iter.Get()
+					resolved = true
+					return nil
+				}
+			}
+		}
+		return errors.New("error deriving publickey from activitystreamsperson")
+	}
+
+	serviceCallback := func(c context.Context, service vocab.ActivityStreamsService) error {
+		if pkProp := service.GetW3IDSecurityV1PublicKey(); pkProp != nil {
+			for iter := pkProp.Begin(); iter != pkProp.End(); iter = iter.Next() {
+				if iter.IsW3IDSecurityV1PublicKey() {
+					pubkey = iter.Get()
+					resolved = true
+					return nil
+				}
+			}
+		}
+		return errors.New("error deriving publickey from activitystreamsservice")
+	}
+
+	applicationCallback := func(c context.Context, app vocab.ActivityStreamsApplication) error {
+		if pkProp := app.GetW3IDSecurityV1PublicKey(); pkProp != nil {
+			for iter := pkProp.Begin(); iter != pkProp.End(); iter = iter.Next() {
+				if iter.IsW3IDSecurityV1PublicKey() {
+					pubkey = iter.Get()
+					resolved = true
+					return nil
+				}
+			}
+		}
+		return errors.New("error deriving publickey from activitystreamsapp")
+	}
+
+	pubkeyCallback := func(c context.Context, pk vocab.W3IDSecurityV1PublicKey) error {
+		pubkey = pk
+		resolved = true
+		return nil
+	}
+
+	if e := ResolveIRI(context.Background(), publicKeyIRI, personCallback, serviceCallback, applicationCallback, pubkeyCallback); e != nil {
+		err = e
+	}
+
+	if err != nil {
+		err = errors.Wrap(err, "error resolving publickey from iri")
+	}
+
+	if !resolved {
+		err = errors.New("error resolving publickey from iri")
+	}
+
+	return pubkey, err
+}
+
 // GetResolvedActorFromIRI will resolve an IRI string to a fully populated actor.
 func GetResolvedActorFromIRI(personOrServiceIRI string) (apmodels.ActivityPubActor, error) {
 	var err error

